src/main/kotlin/org/rust/cargo/project/model/impl/CargoProjectImpl.kt

```diff
@@ -212,6 +213,87 @@ open class CargoProjectsServiceImpl(
             .flatMap { ModuleRootManager.getInstance(it).contentRoots.asSequence() }
             .mapNotNull { it.findChild(RustToolchain.CARGO_TOML) }
 
+    fun updateFeature(
+        cargoProject: CargoProjectImpl,
+        pkg: CargoWorkspace.Package,
+        name: String,
+        enable: Boolean,
+        isDocUnsaved: Boolean
+    ) {
+        if (isDocUnsaved) {
+            runWriteAction { saveAllDocuments() }
+            refreshAllProjects()
+        }
+
+        val userOverriddenFeatures = cargoProject.userOverriddenFeatures
+            .mapValues { (_, v) -> v.toMutableSet() }
+            .toMutableMap()
+        val packageRoot = pkg.rootDirectory.toString()
+        val packageFeatures = userOverriddenFeatures.getOrPut(packageRoot) { hashSetOf() }
+
+        projects.updateSync { projects ->
+            val oldProject = projects.singleOrNull { it.manifest == cargoProject.manifest }
+                ?: return@updateSync projects
+            val workspace = oldProject.workspace ?: return@updateSync projects
+
+            if (enable) {
+                /** When the user enables a feature, all we have to do is add it into [userOverriddenFeatures] */
+                packageFeatures.add(name)
+            } else {
+                /**
+                 * But when disables, we should disable all the features that were enabled automatically due to this one
+                 *
+                 * For example, consider such a case:
+                 * (let [a] mean the feature was enabled automatically, [u] – by the user)
+                 * - [u] foo = []
+                 * - [a] bar = []
+                 * - [u] foobar ["foo", "bar"]
+                 * - [u] baz = ["foo", "bar", "foobar"]
+                 *
+                 * When the user disables the feature `foobar`, the feature `baz` will be disabled transitively.
+                 * But we also should disable `bar` because the user did not enable it. So we finished with this:
+                 * - [u] foo = []
+                 * - [ ] bar = []
+                 * - [ ] foobar ["foo", "bar"]
+                 * - [ ] baz = ["foo", "bar", "foobar"]
+                 */
```

Comment:
Suuuuper drive by comment, but the shape of code here feels wrong to me (which quite probably means that I just don't know the whole context).

Naively, I'd expect something like this

```Kotlin
class Features(
    enabledByTheUser: HashSet<String>,
) {
   val enabled: HashSet<String> get() {
     // computes transitive closure
   }
}
```

That is, it seems like enabling or disabling feature should *just* add it to the set of explicitelly checked features. 

Separately, there should be a thing that computes the full set of features which are activated, based on the set of features which are explicitelly selected.

Reply:
I like this more explicit solution, but it would probably require 2 checkboxes in the UI. One read-only which says whether the feature is enabled [transitively] or not, and one writable checkbox that selects whether the user explicitly wants this feature to be enabled. I think that the solution with 2 checkboxes would generate more predictable behaviour, but it might also confuse users.

Reply:
https://jetbrains.design/intellij/controls/checkbox/#14 ?

Reply:
The states are selected, unselected but enabled and disabled. That seems like a good usecase for a three-state checkbox indeed. When the feature is enabled but not selected, clicking on it should select it and not disable it along with all super-features IMO. It could also simplify the implementation. It depends on who likes which UI design :)